### PR TITLE
Declaration with array assignment

### PR DESCRIPTION
**Pull request by hiker**
Hi,

this fixes lines like "integer a(3) = [1,2,3]". The parser didn't like the [] ('(/ /)' worked), and I also improved handling of quotes in case of assignment of arrays of strings.

### CODE REVIEW

**Comment by hiker** [comment]
Arg, I just realised that I branched this off the "hiker:more-parser-tests" branch, not master :(
I think once you have merged my previous pull request, this one should work without a conflict .. otherwise I'll fix it.

**Comment by arporter** [comment]
@hiker, please could you fix the conflicts in this branch and then I'll review it, thanks!

**Comment by arporter** [comment]
@hiker, is this ready for review now?

**Comment by hiker** [comment]
I think it was kind of ready before, but now it is more ready ;)
Oh, the one line that is not covered in splitline ('continue') is actually executed (put a print statement in front of the continue, and you will get 100% coverage) - must be an issue with the coverage tool.

File: src/fparser/tests/test_splitline.py [hunk 1]
```diff
@@ -0,0 +1,139 @@
+# Modified work Copyright (c) 2017 Science and Technology Facilities Council
```
Comment (arporter) [inline src/fparser/tests/test_splitline.py]: New test file containing tests previously included in splitline source (and therefore presumably not run by py.test). Great.

File: src/fparser/tests/test_splitline.py [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+         '(/"a[),", ",b,[(", "c,][)("/)']
+    assert splitparen('integer a(3) = ["a", "b", "c"]') == \
+        ['integer a', '(3)', ' = ', '["a", "b", "c"]']
+    assert splitparen(
+        'character(len=40) :: a(3) = ["a[),", ",b,[(", "c,][)("]') == \
+        ['character', '(len=40)', ' :: a', '(3)', ' = ',
+         '["a[),", ",b,[(", "c,][)("]']
+    # pylint: disable=anomalous-backslash-in-string
+    result = splitparen('a(1),b\\((2,3),c\\\((1)),c"("')
+    expected = ['a', '(1)', ',b\(', '(2,3)', ',c\\\\', '((1))', ',c"("']
```
Comment (arporter) [inline src/fparser/tests/test_splitline.py]: I can see that you've brought this test in unchanged from how it was before and therefore you might not know the answer to this, but... why does e.g. "c\\\" become "c\\\\" while "b\\" becomes "b\"? My Fortran knowledge does not stretch to this situation. Maybe yours or that of @rupertford does?

Comment (hiker) [reply src/fparser/tests/test_splitline.py]: It's not a Fortran, but a python issue - I've modified the tests somewhat to be easier to understand. The previous tests made use of the fact that python would not change a '\(' - this stays a two character strings. So e.g.: 'c\\\(' == 'c\\\\(' in python:
```
>>> print 'b\(' == 'b\\(', 'c\\\(' == 'c\\\\('
True True
```
Hope this is easier to understand now - the actual test case was not changed, but the way the string was written for the test case was improved.

**Comment by hiker** [comment]
OK, as far as I can see all concerns were addressed.

File: src/fparser/splitline.py [hunk 3]
```diff
@@ -176,313 +182,122 @@ def splitquote(line, stopchar=None, lower=False, quotechars = '"\''):
...
+def splitparen(line, paren_open="([", paren_close=")]"):
     """
-    Fast LineSplitterParen.
+    Splits a line into top-level parenthesis and not-parenthesised
+    parts. E.g.: "a( (1+2)*3) = b(x)" becomes:
+    ["a", "( (1+2)*3)", " = b", "(x)"]
+    Line: the string to split.
+    paren_open/close: The characters that define the parentheses.
+    They must be matched in order: paren_open[x] is closed by
+    paren_close[x].
```
[the remark this answers is not in the document]
Comment (hiker) [reply src/fparser/splitline.py]: Done ... what's wrong with doxygen?? *sigh* Another standard to learn ;) 
(and that was a rhetorical question ;) )

File: src/fparser/splitline.py [hunk 4]
```diff
@@ -176,313 +182,122 @@ def splitquote(line, stopchar=None, lower=False, quotechars = '"\''):
...
+    items = []   # Result list
+    i = 0        # Index of current character
+    num_backslashes = 0   # Counts consecutivre "\" characters
+    # Empty if outside quotes, or set to the starting (and therefore
+    # also the ending) quote character while reading text inside quotes.
+    inside_quotes_char = ""
+    start = 0    # Index of start of current part.
+    stack = []   # Stack keeping track of required closing brackets
+
+    i = -1
```
Comment (arporter) [inline src/fparser/splitline.py]: This line is no-longer needed.

File: src/fparser/splitline.py [hunk 5]
```diff
@@ -176,313 +182,122 @@ def splitquote(line, stopchar=None, lower=False, quotechars = '"\''):
...
+    for i, char in enumerate(line):
+        char = line[i]
```
Comment (arporter) [inline src/fparser/splitline.py]: I don't think this line is needed either.

File: src/fparser/splitline.py [hunk 6]
```diff
@@ -176,313 +182,122 @@ def splitquote(line, stopchar=None, lower=False, quotechars = '"\''):
...
+    i = 0        # Index of current character
+    num_backslashes = 0   # Counts consecutivre "\" characters
+    # Empty if outside quotes, or set to the starting (and therefore
+    # also the ending) quote character while reading text inside quotes.
+    inside_quotes_char = ""
+    start = 0    # Index of start of current part.
+    stack = []   # Stack keeping track of required closing brackets
+
+    i = -1
+    for i, char in enumerate(line):
```
Comment (arporter) [inline src/fparser/splitline.py]: Please could you change `i` to e.g. `idx` (pep8)

File: src/fparser/tests/test_splitline.py [hunk 7]
```diff
@@ -0,0 +1,139 @@
...
+# ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+# DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+# SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+# CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+# LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+# OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH
+# DAMAGE.
+
+# Original author: Pearu Peterson <pearu@cens.ioc.ee>
+# First version created: May 2006
```
Comment (arporter) [inline src/fparser/tests/test_splitline.py]: Please add yourself to the list, e.g.:

    # Modified by J. Henrichs, affiliation

File: src/fparser/splitline.py [hunk 8]
```diff
@@ -176,313 +182,122 @@ def splitquote(line, stopchar=None, lower=False, quotechars = '"\''):
...
 
-                if char=='\\':
-                    nofslashes += 1
-                else:
-                    nofslashes = 0
+    assert len(paren_open) == len(paren_close)
+
+    items = []   # Result list
+    i = 0        # Index of current character
+    num_backslashes = 0   # Counts consecutivre "\" characters
```
Comment (arporter) [inline src/fparser/splitline.py]: s/consecutivre/consecutive/

File: src/fparser/splitline.py [hunk 9]
```diff
@@ -176,313 +182,122 @@ def splitquote(line, stopchar=None, lower=False, quotechars = '"\''):
...
     """
-    Fast LineSplitterParen.
+    Splits a line into top-level parenthesis and not-parenthesised
+    parts. E.g.: "a( (1+2)*3) = b(x)" becomes:
+    ["a", "( (1+2)*3)", " = b", "(x)"]
+    :param str line: the string to split.
+    :param str paren_open: The characters that define an open parentheses.
+    :param str paren_close: The characters that define a closing parentheses.
+    The paren_open and paren_close strings must be matched in order:
+    paren_open[x] is closed by paren_close[x].
```
Comment (arporter) [inline src/fparser/splitline.py]: Great but please also add:

    :return: List of parenthesised and not-parenthesised parts
    :rtype: list of str

File: src/fparser/splitline.py [hunk 10]
```diff
@@ -72,25 +72,28 @@
 -----
```
Comment (arporter) [inline src/fparser/splitline.py]: Please add yourself as an author of this file.

File: src/fparser/splitline.py (hunk 6 above)
Comment (hiker) [reply src/fparser/splitline.py]: Neither pep8 nor pylint complained about this, and I personally prefer using i as loop index - but I fixed it.

**Comment by hiker** [comment]
Hope I got the (c) at the right place and right format - just change it if you prefer it somewhere else.

**Pull request merged**